src/jsonrpc.js

```diff
@@ -17,14 +33,85 @@ export class JsonRpc {
   }
 
   /**
-   * Scanning all the blocks fromBlock and toBlock 
-   * which are internal variables passed via the constructor
+   * 
+   * This will return the final data structure 
+   * for the transaction resopnse
+   * @param string tx 
+   * @param Object receipt
+   * @param Array logs
+   * @return object
+   */
+  static getTransactionFormat(transaction, receipt, logs) {
+    const receiptResult = receipt;
+
+    receiptResult.logs = logs;
+    return { ...transaction, ...receiptResult };
+  }
+
+  /**
+   * Async function that gets the transaction and transaction receipt 
+   * then get the logs out of the receipt transaction then execute the callback function  
+   * @param string txn
+   */
+  async _scanTransaction(txn) {
+    const txnReceipts = await toPromise(this.web3Instance.getTransactionReceipt)(txn.hash);
+    const logs = txnReceipts.logs.filter(log => isInArray(this.addresses, log.address));
+
+    /**
+     * This case should never happend, because we are cheching 
+     * the smart contract not the normal addresses
+     */
+    if (txn.from && isInArray(this.addresses, txn.from.toLowerCase())) {
+      throw new Error('Smart contract is shouldn\'t be in from');
+    }
+
+    // If the smart contract recieved transaction or there's logs execute the callback function 
+    if ((txn.to && isInArray(this.addresses, txn.to.toLowerCase())) || logs.length > 0) {
+      const transactionResult = JsonRpc.getTransactionFormat(txn, txnReceipts, logs);
+      if (this.callback) { this.callback(transactionResult); }
+    }
+  }
+
+  /**
+   * This function handles the transactions that exist in one block
+   *  and puts them into an array of promises, then executes them.
+   * @param object block 
+   */
+  async _scanBlockCallback(block) {
+    if (block.transactions) {
```

Comment:
I think there should be an extra check for blocks here `block` might be `undefined`
Additionally, I think this function should throw if block is undefined that means there is a big problem with our RPC connection

Reply:
👍